pkg/experiment/ingester/segment_test.go

```diff
@@ -227,20 +221,18 @@ func TestBusyIngestLoop(t *testing.T) {
 }
 
 func TestDLQFail(t *testing.T) {
-
 	l := testutil.NewLogger(t)
-	bucket := mockBucket{upload: func(ctx context.Context, name string, r io.Reader) error {
-		if strings.HasSuffix(name, pathBlock) {
-			return nil
-		}
-		assert.Contains(t, name, pathDLQ)
-		return fmt.Errorf("mock upload DLQ error")
-	}}
-	client := &metastoreClientMock{
-		addBlock: func(ctx context.Context, in *metastorev1.AddBlockRequest, opts ...grpc.CallOption) (*metastorev1.AddBlockResponse, error) {
-			return nil, fmt.Errorf("metastore unavailable")
-		},
-	}
+	bucket := mockobjstore.NewMockBucket(t)
+	bucket.On("Upload", mock.Anything, mock.MatchedBy(func(name string) bool {
+		return strings.HasSuffix(name, pathBlock)
+	}), mock.Anything).Return(nil)
+	bucket.On("Upload", mock.Anything, mock.MatchedBy(func(name string) bool {
+		return strings.Contains(name, pathDLQ)
+	}), mock.Anything).Return(fmt.Errorf("mock upload DLQ error"))
+	client := mockmetastorev1.NewMockMetastoreServiceClient(t)
+	client.On("AddBlock", mock.Anything, mock.Anything, mock.Anything).
+		Return(nil, fmt.Errorf("mock add block error"))
+
```

Comment:
You may also want to specify whether each method (and how many times) was called and `AssertExpectations` in the end. Otherwise, if a method has not been called, the test will pass (unless you assert side-effects, which you do)